kernel/src/syscall/getdents64.rs

```diff
@@ -102,12 +98,32 @@ impl<'a, T: DirentSerializer> DirentBufferReader<'a, T> {
 
 impl<T: DirentSerializer> DirentVisitor for DirentBufferReader<'_, T> {
     fn visit(&mut self, name: &str, ino: u64, type_: InodeType, offset: usize) -> Result<()> {
+        /// The maximum bytes in a file name is 255 in fs/utils/mod.rs
+        const MAX_BUFFER_SIZE: usize = 512;
         let dirent_serializer = T::new(ino, offset as u64, type_, CString::new(name)?);
-        if self.read_len >= self.buffer.len() {
-            return_errno_with_message!(Errno::EINVAL, "buffer is too small");
+        let len = dirent_serializer.len();
+        if self.writer.avail() < len {
+            return_errno_with_message!(
+                Errno::EINVAL,
+                "user buffer has insufficient space for directory entry"
+            );
+        }
+
+        let mut buffer = [0u8; MAX_BUFFER_SIZE];
+        if len > buffer.len() {
+            return_errno_with_message!(
+                Errno::EINVAL,
+                "directory entry exceeds maximum supported size"
+            );
```

Comment:
Move the validation logic from lines 113–118 to before line 112. This ensures we don't allocate the buffer if the check fails.

```rust
if len > MAX_BUFFER_LEN {
    // ......
}
```

Reply:
fixed